vroom/pluginextensions.vroom

```diff
@@ -0,0 +1,200 @@
+  :let g:repo = maktaba#path#Join([g:thisdir, 'fakeplugins'])
+  :let g:pluginpath = maktaba#path#Join([g:repo, 'myplugin'])
+  :let g:plugin = maktaba#plugin#Install(g:pluginpath)
+
+Plugins have an individual 'extension registry' that holds all the
+extensions registered for use by that plugin.
+
+A plugin obtain its own extension registry via the existing plugin object.
+
+  :let g:registry = g:plugin.GetExtensionRegistry()
+
+A plugin's extension registry contains functions to modify and query the
+list of extensions registered for that plugin, and to define what makes a
+valid extension.
+
+Extensions are arbitrary dictionaries, so a plugin providing an extension
+point can optionally (and should) register a validator that will check that
+a provided extension is acceptable, via the extension registry's
+SetValidator() function.
+
+  :function! MyFormatterValidator(extension) abort
+  :  if !has_key(a:extension, 'filetype')
+  :    throw maktaba#error#BadValue('No filetype.')
+  :  endif
+  :  if !has_key(a:extension, 'command')
+  :    throw maktaba#error#BadValue('No command.')
+  :  endif
+  :endfunction
+  :call g:registry.SetValidator('MyFormatterValidator')
+
+SetValidator() accepts a string, funcref, or Maktaba function dictionary,
+just like maktaba#function#Call().
+
+
+Now, when a second plugin comes to register an extension for the above
+plugin to use, it must also get access to that plugin's extension registry.
+
+Note that the second plugin cannot use the above mechanism to do so, as it
+would require that plugin to load the first plugin directly, which would
+only be permitted if that was a library plugin.
+
+Instead, Maktaba allows plugins to register extensions against the name of
+the original plugin. This does not require that the original plugin is
+installed (or even that it will be installed - it is not an error to
+register extensions against plugins that the user is not using).
+
+The following would normally be run from a separate plugin during loading,
+but for the purposes of exposition, we're going to make the calls directly.
+
+This time, the extension registry is obtained from a call to
+maktaba#extension#GetRegistry().
+
+  :let g:registry = maktaba#extension#GetRegistry('myplugin')
+
+Once the plugin has an extension registry for the target plugin, it can add
+extensions using the AddExtension() function.
+
+  :let g:foo_extension = {'filetype': 'foo', 'command': 'foo-format'}
+  :call g:registry.AddExtension(g:foo_extension)
+
+If the original plugin has registered a validator (as is the case here), the
+validator will be called to verify the extension.  Any errors returned from
+the validator will be shouted to the user (and will prevent the extension
+from being registered), but will not be propagated to the caller.
+
+TODO: would it make sense for AddExtension() to return whether it succeed,
+then?
+
+TODO: should we define what exception the validator should throw (or even
```

Comment:
I think it's fine. I considered only catching `ERROR(.*)` and letting unexpected code errors like "Invalid expression" bubble up normally, but I can't really see that being a big deal, either.